internal/querynodev2/delegator/delegator.go

```diff
@@ -684,12 +705,164 @@ func NewShardDelegator(ctx context.Context, collectionID UniqueID, replicaID Uni
 		loader:          loader,
 		factory:         factory,
 		queryHook:       queryHook,
+		vectorField:     vectorField,
 	}
 	m := sync.Mutex{}
 	sd.tsCond = sync.NewCond(&m)
 	if sd.lifetime.Add(lifetime.NotStopped) == nil {
 		go sd.watchTSafe()
 	}
+
 	log.Info("finish build new shardDelegator")
 	return sd, nil
 }
+
+func getVectorFieldFromSchema(schema *schemapb.CollectionSchema) (*VectorField, error) {
+	vecFieldSchema, err := typeutil.GetVectorFieldSchema(schema)
+	if err != nil {
+		return nil, err
+	}
+	dimStr, err := funcutil.GetAttrByKeyFromRepeatedKV(common.DimKey, vecFieldSchema.GetTypeParams())
+	if err != nil {
+		return nil, err
+	}
+	dimValue, err := strconv.ParseInt(dimStr, 10, 32)
+	if err != nil {
+		return nil, err
+	}
+
+	vectorField := VectorField{
+		dim:       dimValue,
+		dataType:  vecFieldSchema.GetDataType(),
+		fieldName: vecFieldSchema.GetName(),
+	}
+	return &vectorField, nil
+}
+
+// OptimizeSearchBasedOnClustering optimize SearchRequest based on segments' clustering info.
+// Basic rule: calculate distance between search vector and clustering centroid of each segment,
+// only search on top x% nearest segment
+// Todo support call external hook
+func (sd *shardDelegator) OptimizeSearchBasedOnClustering(req *querypb.SearchRequest, sealeds []SnapshotItem) (*querypb.SearchRequest, []SnapshotItem) {
+	log := log.With(zap.Int64("searchRequestID", req.GetReq().GetReqID()))
+	if !paramtable.Get().QueryNodeCfg.EnableSearchBasedOnClustering.GetAsBool() {
+		log.Debug("skip OptimizeSearchBasedOnClustering by system config")
+		return req, sealeds
+	}
+	if req.GetReq().GetClusteringOptions().GetFilterRatio() >= 1 {
+		log.Debug("skip OptimizeSearchBasedOnClustering by user config")
+		return req, sealeds
+	}
+	metricType := req.GetReq().GetMetricType()
+	topK := req.GetReq().GetTopk()
+	filterRatio := req.GetReq().GetClusteringOptions().GetFilterRatio()
```

Comment:
We should probably make this a collection level config?

There is no need to config in search request

Reply:
EnableSearchBasedOnClustering could also be a collection level configs

Reply:
add we can add a check to only enable this feature for float32 vectors